R/generate_retro_triangles.R

```diff
@@ -0,0 +1,108 @@
+#' Generate retrospective reporting triangles
+#'
+#' This function ingests a reporting triangle and the number of retrospective
+#'   reporting triangles we want to create, and iteratively generates the
+#'   reporting triangle that would have been available as of the maximum
+#'   reference time, and iteratively working backwards for
+#'   `n_history_uncertainty` snapshots
+#'
+#' @param triangle Matrix of the reporting triangle
+#'   to be used to generate retrospective nowcasts, with rows representing the
+#'   time points of reference and columns representing the delays
+#' @param n_triangles Integer indicating the number of retrospective
+#'   reporting triangles to be generated, always starting from the most
+#'   recent reference time. Default is  only generate truncated matrices
+#'   that have sufficient number of rows to generate a nowcast from, though
+#'   any number can be specified.
+#' @returns A list of `n_triangle` retrospective reporting triangle matrices
+#'   with as many rows as available given the truncation, and the same number
+#'   of columns as `triangle`
+#' @export
+#' @examples
+#' triangle <- matrix(
+#'   c(
+#'     65, 46, 21, 7,
+#'     70, 40, 20, 5,
+#'     80, 50, 10, 10,
+#'     100, 40, 31, 20,
+#'     95, 45, 21, NA,
+#'     82, 42, NA, NA,
+#'     70, NA, NA, NA
+#'   ),
+#'   nrow = 7,
+#'   byrow = TRUE
+#' )
+#'
+#' retro_rts <- generate_retro_triangles(
+#'   triangle = triangle,
+#'   n_triangles = 2
+#' )
+#' print(retro_rts[[1]])
+#' print(retro_rts[[2]])
+generate_retro_triangles <- function(
+    triangle,
+    n_triangles = nrow(triangle) - ncol(triangle)) {
+  .validate_triangle(triangle)
+  if (n_triangles > (nrow(triangle) - ncol(triangle))) {
+    cli::cli_warn(
+      message = c(
+        "Not all of the triangles generated will contain sufficient ",
+        "observations to generate a nowcast from"
+      )
+    )
+  }
+
+  # Will be able to remove this step if we require NAs in the bottom right
+  # of the triangle
+  matr_observed <- .replace_lower_right_with_NA(triangle)
+
+  results <- lapply(seq_len(n_triangles),
+    generate_retro_triangle,
+    matr_observed = matr_observed
+  )
+
+  return(results)
+}
+
+
+#' Get a single retrospective triangle
+#'
+#' This function takes in a integer `t` and a reporting triangle and generates
+#'  the reporting triangle that would have been observed as of `t` units
+#'  earlier (starting from the bottom of the reporting triangle)
+#'
+#' @param t Integer indicating the number of days prior to generate the
+#'  retrospective reporting triangle for
+#' @param matr_observed Matrix of the incomplete reporting triangle
```

Comment:
Is it enforced to be incomplete? If so I think we might want a new issue to remove this constraint in the future as it seems like a constraint that isn't needed

Reply:
I will edit the documentation here, it does not need to be incomplete-- the conversion from truncated matrix to reporting triangle happens within this function when `.replace_bottom_right_with_NA()` is called.